src/chemnlp/data_val/config.py

```diff
@@ -63,9 +66,41 @@ class TrainPipelineConfig(BaseModel):
     trainer: TrainerConfig
     wandb: WandbConfig
 
+    def update(
+        self, config_changes: Dict[str, Dict[str, Any]]
+    ) -> "TrainPipelineConfig":
+        """Update training configuration"""
+        for config_key, parameter_changes in config_changes.items():
+            # top level config classes
+            config_attr = getattr(self, config_key)
+            for param_key, param_value in parameter_changes.items():
+                # second level configuration parameters
+                setattr(config_attr, param_key, param_value)
+        return self
+
 
 class DataMixingConfig(BaseModel):
     data_paths: List[str]
     data_proportions: List[float]
     save_path: str
     stopping_strategy: str
+
+
+class GridSearch(BaseModel):
+    """Grid search options for TrainPipelineConfig elements"""
+
+    data: Optional[DictofLists] = {}
+    model: Optional[DictofLists] = {}
+    prompt_tuning: Optional[DictofLists] = {}
+    trainer: Optional[DictofLists] = {}
+    wandb: Optional[DictofLists] = {}
+
+
+def _get_all_combinations(d: Dict):
+    """Generate all possible hyperparameter combinations"""
+    keys, values = d.keys(), d.values()
+    values_choices = (
+        _get_all_combinations(v) if isinstance(v, dict) else v for v in values
+    )
+    for comb in itertools.product(*values_choices):
+        yield dict(zip(keys, comb))
```

Comment:
I don't know if this function would be better suited to go somewhere else so this file contains only configs?
Maybe it could be inside experiments/scripts/run_grid_search.py since currently it will be used only there? Totally up to you though :)